session.py

```diff
@@ -31,13 +28,25 @@ def __init__(self):
 
 class BidiSessionSignature(Signature):
     """
-    - Start with inbound.active raised: "new request"
-    - Wait for outbound.active to be raised before inputting data
-    - Data flows
-    - Wait until outbound data flushed before deasserting session_active
-    - Early termination (e.g. 400): flush inbound data until inbound.session_active deasserted
-    - Both zero --> reset state
-
+    A bidirectional session-oriented data stream.
+    A pair of SessionSignatures that control inbound and outbound data,
+    and provide a bidirectional handshake for starting a new session.
+
+    The sequence of usage is as follows:
+
+    - At reset, inbound.active and outbound.active are both zero.
+    - A new session starts with inbound.active raised
+    - The session manager wait for outbound.active to be raised before
+      passing data into inbound.data. inbound.data and outbound.data
+      perform flow control as usual
+    - Data flows for a while
+    - At the end of a session, inbound.active OR outbound.active deasserts,
+      after all data has been consumed (i.e. data.valid goes low for
+      the final time)
+    - The still-active end must continue consuming data until
```

Comment:
IMO, might be better to be more specific:

- Data may be pushed at any time while the respective .active signal is asserted--i.e., while `inbound.active` is true, `inbound.data.valid` may become true, and similarly for `outbound.active` and `outbound.data.valid`. The consumer side of these streams must acknowledge this data by setting the appropriate `ready` signal.

Or something like that?

Reply:
I'd like for the semantics of this to match TCP (for future use), which is a little more restrictive- your statement is true *once a session is established*, but we don't want to receive/acknowledge/enqueue bytes until both `.active` lines have asserted:

- Data can flow once the session is established, i.e. `inbound.data.valid` will not become ready until both `active` signals are asserted.
- Once the session is established, new data may appear (`.valid` asserted) as long as the corresponding `.active` signal is high. Once an `.active` signal goes low, the corresponding `.valid` signal will not re-assert until a new session is established.



Does that make sense? Included in f2985bf

Reply:
The statement I thought you were trying to make was about the "ready" signal? Doesn't look like that's captured anymore.

Reply:
Not trying to say anything about `.ready` in particular -- since that's really just `stream.Signature`. Rephrased a little further, is this clearer?

Reply:
I think so. I guess I'm conflating "continue consuming data" with the implementation detail of ready going high.

Resoved.